Add RareTerms aggregation

polyfractal opened the pull request:
This adds a `rare_terms` aggregation.  It is an aggregation designed to identify the long-tail of keywords, e.g. terms that are "rare" or have low doc counts.

This aggregation is designed to be more memory efficient than the alternative, which is setting a terms aggregation to `size: MAX_LONG` (or worse, ordering a terms agg by count ascending, which has unbounded error).

This aggregation works by maintaining a map of terms that have been seen. A counter associated with each value is incremented when we see the term again.  If the counter surpasses a predefined threshold, the term is removed from the map and inserted into a bloom filter.  If a future term is found in the bloom filter we assume it was previously removed from the map and is "common".

The map keys are the "rare" terms after collection is done. 

## Outstanding issues

- Unclear how we should default the bloom filter.
- Do we expose the bloom filter params to the user, or try to pick a one-size fits all?  I think we probably need to expose some settings but right now nothing is configurable
- What's the max `max_doc_count` that we allow?  Currently set to 10 but I think that's probably too low.  It's mainly another safety mechanism, the max buckets limit will still trigger too.  It might not make sense to even have a max here, since it's pretty data-dependent.
- No global ordinal support for strings.  The PR was already huge so I think this should be done in a followup?
- A few misc items questions in `//TODO review` comments
- I don't have any "big" integration tests, just the yaml tests.  Should I add a QA test or something that tests this on a few thousand docs?

Closes #20586 (finally!)

@andyb-elastic @not-napoleon tagged you both as reviewers in case your interested but no pressure if not, or too busy :)

Also /cc @clintongormley

Commit:
Add RareTerms aggregation

This adds a `rare_terms` aggregation.  It is an aggregation designed
to identify the long-tail of keywords, e.g. terms that are "rare" or
have low doc counts.

This aggregation is designed to be more memory efficient than the
alternative, which is setting a terms aggregation to size: LONG_MAX
(or worse, ordering a terms agg by count ascending, which has
unbounded error).

This aggregation works by maintaining a map of terms that have
been seen. A counter associated with each value is incremented
when we see the term again.  If the counter surpasses a predefined
threshold, the term is removed from the map and inserted into a bloom
filter.  If a future term is found in the bloom filter we assume it
was previously removed from the map and is "common".

The map keys are the "rare" terms after collection is done.

elasticmachine commented:
Pinging @elastic/es-search-aggs

polyfractal commented on server/src/main/java/org/elasticsearch/common/util/BloomFilter.java:
```diff
@@ -0,0 +1,381 @@
+/**
+ * A bloom filter. Inspired by Guava bloom filter implementation though with some optimizations.
+ */
+public class BloomFilter implements Writeable, Releasable {
```
This class was resurrected from the depths of git.  This BloomFilter used to be used by ES elsewhere (doc IDs I think?), but I just realized none of the tests made it through the resurrection.

I'll start looking for those tests, or add my own.

This class had a lot of extra cruft that wasn't needed anymore (string configuration parsing, factories, multiple hashing versions, etc) so I tried to simplify it where possible.

clintongormley commented:
/cc @tsg

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractStringTermsAggregator.java:
```diff
@@ -33,7 +33,7 @@
 import static java.util.Collections.emptyList;
 
-abstract class AbstractStringTermsAggregator extends TermsAggregator {
+public abstract class AbstractStringTermsAggregator extends TermsAggregator {
```
Why is this needed? I don't see this class being use anywhere extra but maybe I'm missing something?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleRareTerms.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Result of the RareTerms aggregation when the field is some kind of decimal number like a float, double, or distance.
+ */
+public class DoubleRareTerms extends InternalMappedRareTerms<DoubleRareTerms, DoubleTerms.Bucket> {
```
I'd love for us to not implement this one since doubles and floats are likely to have an extremely long tail and therefore make using this aggregation on them untenable

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/DoubleRareTermsAggregator.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DoubleRareTermsAggregator extends LongRareTermsAggregator {
```
I'd love for us to not implement this one since doubles and floats are likely to have an extremely long tail and therefore make using this aggregation on them untenable

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedRareTerms.java:
```diff
@@ -0,0 +1,132 @@
+            }
+
+            if (bloomFilter == null) {
+                bloomFilter = ((InternalMappedRareTerms)aggregation).bloom;
```
I think we should always merge into a new bloom filter. We have had issues in the past with reusing state from the shard responses so we tend to create new state for the reduce result now.

polyfractal replied:
++ will do

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedRareTerms.java:
```diff
@@ -0,0 +1,132 @@
+            // up in a later incremental reduction and looks "rare" even though it isn't.
+            if (reduceContext.isFinalReduce() == false || (b.getDocCount() <= maxDocCount && containsTerm(bloom, b) == false)) {
+                rare.add(b);
+            }
```
For the terms which merge and end up above the doc count limit do we not need to add them to the bloom filter? I think this is especially important for the case where there are multiple reduce phases (incremental reduce)

polyfractal replied:
Hmm, I think it's protected from that right now because we only "prune" the list if it's the final reduction, so we know there are no more terms and all same-term buckets are combined.

But that was added because there were issues with incremental reduction, and I think your proposal (add to bloom) would fix the issue without having to wait until the final reduction like I'm doing right now.

Will investigate, thanks :)

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java:
```diff
@@ -0,0 +1,204 @@
+    static final BucketOrder ORDER = BucketOrder.compound(BucketOrder.count(true), BucketOrder.key(true)); // sort by count ascending
+
+    //TODO review question: is LongLong map ok?
+    protected LongLongHashMap map;
```
Is this needed? Since we are calling `collectBucket` and `collectExistingBucket` as we go through collection, can you not ask the bucket for its doc count rather than duplicating it in a separate data structure?

colings86 replied:
I'm still not sure about this because as far as I can see there are the following scenarios when collecting a document:
1. The value in the document is new - we are good because we just add the document to the LongHash and the bucket count is stored as 1.
2. The value in the document has been seen before but is still considered rare (has not hit the threshold doc count) - In this case we are good because we can ask for the doc count using ` bucketDocCount(long bucketOrd)` to determine whether we hit the threshold since the bucketOrd get be found from `bucketOrds`
3. The value in the document has hit the threshold already but the we haven't yet run the merge to delete it from bucketOrds etc. - In this case we will check the filter, determine that the value is not rare and not need to consult the buckets at all
4. The value in the document has hit the threshold already and we have since merged it out - In this case we check the filter, determine that the value is not rare and not need to consult the buckets at all

When we do the merge we recreate all the data structures for the buckets anyway (bucketOrds, doc counts and docDeltas, etc.) so I don't think we need to be able to delete single entries from the bucketOrds? I'm probably missing something here but I'm not sure what?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java:
```diff
@@ -0,0 +1,204 @@
+                                        long bucketOrdinal = bucketOrds.add(val);
+                                        if (bucketOrdinal < 0) { // already seen
+                                            bucketOrdinal = - 1 - bucketOrdinal;
+                                            collectExistingBucket(subCollectors, docId, bucketOrdinal);
```
Would getting here not indicate a bug since we have determined the `termCount` to be `0` already so we should not have seen the bucket ordinal before?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java:
```diff
@@ -0,0 +1,204 @@
+                                        if (termCount < maxDocCount) {
+                                            // TODO if we only need maxDocCount==1, we could specialize
+                                            // and use a bitset instead of a counter scheme
+                                            map.put(val, termCount + 1);
```
If we are incrementing the counter here do we not also need to collect the document in the existing bucket?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java:
```diff
@@ -0,0 +1,204 @@
+                    newBucketOrd = newBucketOrds.add(oldKey);
+                } else {
+                    // Make a note when one of the ords has been deleted
+                    hasDeletedEntry = true;
```
To make sure this GC is working correctly I wonder if it's worth having a counter here and then checking the counter value is the same as the numDeleted that we expect at the end of this for loop? Another option would be to initialise the variable to `numDeleted` and decrement it here ensuring it reaches 0.

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java:
```diff
@@ -0,0 +1,204 @@
+                }
+            }
+        }
+        bucketOrds = newBucketOrds;
```
Do we not also need to reset `numDeleted` here?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedRareTerms.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
```
Should we call `consumeBucketsAndMaybeBreak` in the reduce in case the number of rare terms across all the shards reaches the limit?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/RareTermsAggregatorFactory.java:
```diff
@@ -0,0 +1,165 @@
+            ExecutionMode execution = ExecutionMode.MAP; //TODO global ords not implemented yet, only supports "map"
+
+            DocValueFormat format = config.format();
+            if ((includeExclude != null) && (includeExclude.isRegexBased()) && format != DocValueFormat.RAW) {
```
I think the `DocValueFormat.RAW` check is being used to determine that the field used is a string field. But I see a few issues here (unless I'm misunderstanding what this is doing):
* Users can apply custom formats to non-string fields
* The `valuesSource` has already been checked above to be a ValuesSource.Bytes so this can only be a string field here?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringRareTermsAggregator.java:
```diff
@@ -0,0 +1,213 @@
+            private long numDeleted = 0;
+
+            @Override
+            public void collect(int docId, long bucket) throws IOException {
```
Same comments apply as from `LongRareTermsAggregator` above. Also since this logical is almost the same in three places does it make sense to extract it to something common so we can fix it in one place and apply it to all implementations?

polyfractal replied:
I tried hard to refactor collect() and gcDeletedEntries() into one place... and it's just not possible.  There are too many differences between longs and BytesRef.  Map get/set, ordinals, hashing, doc values, etc are all different and there aren't any shared types that allow it to be resolve easily :(

colings86 reviewed (commented):
@polyfractal I left some comments

jpountz reviewed (commented):
I skimmed through the patch. The general idea of how this works makes sense to me, here are some questions:
 - Do we need a `shard_size` parameter? There could be millions of values that have a doc_count of 1 on each shard? And maybe a `size` parameter as well in case hundreds of shards are queried? I usually don't like adding parameters but I'm afraid that this aggregation might be hard to use without those?
 - Maybe we could try to be smarter with the bloom filter and start with a set<integer> that contains hashes that we only upgrade to a lossy bloom filter when it starts using more memory than the equivalent bloom filter.
 - We should somehow register memory that we allocate for the bloom filter and other data structures to the circuit breakers.
 - Do we need to support sub aggregations? It adds quite some complexity. Also compared to terms aggs a lot of terms might be pruned on the coordinating node because they exist on other shards as well, which might require to increase the shard size which in-turn makes sub aggregations even heavier.
 - I'm not convinced sharing the hierarchy with terms aggregations helps? It might even make it harder to do changes to the terms aggregation in the future?

polyfractal commented on docs/reference/aggregations/bucket/rare-terms-aggregation.asciidoc:
```diff
@@ -0,0 +1,295 @@
+WARNING: When aggregating on multiple indices the type of the aggregated field may not be the same in all indices.
+Some types are compatible with each other (`integer` and `long` or `float` and `double`) but when the types are a mix
+of decimal and non-decimal number the terms aggregation will promote the non-decimal numbers to decimal numbers.
+This can result in a loss of precision in the bucket values.
```
++ Seems reasonable to me.  Would cut down some of the complexity of the agg too, which is a nice perk  :)

polyfractal commented on server/src/main/java/org/elasticsearch/common/util/BloomFilter.java:
```diff
@@ -0,0 +1,381 @@
+ */
+public class BloomFilter implements Writeable, Releasable {
+
+    // Some numbers:
```
@jpountz do you happen to know, or know who would?  This class was taken from the old BloomFilter that I think was used for UUID lookups on segments.  

These numbers used to correlate to the string that was passed in the config, and I _think_ they are in the format 
```
<expected insertions> = <false positive probability : <bloom size> , <num hashes>
```

jpountz replied:
I don't know for sure but would assume the same format indeed.

polyfractal replied:
:+1: thanks.  I'll reformat and tidy up the comment so it makes a bit more sense in the current code

polyfractal commented:
Thanks for the reviews @colings86 @jpountz.  Will try to get to them this week.

> Do we need a shard_size parameter? There could be millions of values that have a doc_count of 1 on each shard? And maybe a size parameter as well in case hundreds of shards are queried? I usually don't like adding parameters but I'm afraid that this aggregation might be hard to use without those?

I agree this is an issue... but doesn't adding `size`/`shard_size` open the agg back up to the type of sharding errors we're trying to avoid?  The shard errors + bloom filter errors may be difficult for a user to understand, leading to nearly as bad of results as a terms agg.  We'd also have to add sorting back (at least by `_term` asc/desc) so that the user could choose which part of the list is truncated.

Perhaps we make it an all-or-nothing agg, and spell out the ramifications in the docs clearly?  E.g. track as we add to the map of potentially-rare terms, and if we ever breach the `max_buckets` threshold we just terminate the aggregation?  So if the user wants accurate rare terms (within the bounds of bloom error), they need to ensure they have configured their `max_buckets` appropriately?

> Maybe we could try to be smarter with the bloom filter and start with a set that contains hashes that we only upgrade to a lossy bloom filter when it starts using more memory than the equivalent bloom filter.
> We should somehow register memory that we allocate for the bloom filter and other data structures to the circuit breakers.
> I'm not convinced sharing the hierarchy with terms aggregations helps? It might even make it harder to do changes to the terms aggregation in the future?

++ Will look into these.  I think they make sense, and if we don't mind a bit of extra c/p decoupling from the terms agg would simplify a few things elsewhere.

> Do we need to support sub aggregations? It adds quite some complexity. Also compared to terms aggs a lot of terms might be pruned on the coordinating node because they exist on other shards as well, which might require to increase the shard size which in-turn makes sub aggregations even heavier.

I'm not sure... I feel like users may want to run sub-aggs on their rare terms.  But not positive.  @clintongormley @tsg do you have any thoughts on this?

colings86 commented:
@elastic/es-analytics-geo

Commit:
Make bloom filter less bad (cleanup legacy cruft)

Commit:
review cleanup

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Add bloom and map to CB

Commit:
Refactor bloom to track exact set, add tests

Commit:
Tweak CB

Commit:
Decouple from terms agg

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Test tweaks, comments

Commit:
Fix doc tests, add back final on InternalMappedTerms#writeTermTypeInfoTo

polyfractal commented:
Got some more work done!

- Decoupled from Terms agg classes.  Needed a decent amount of c/p, but it did simplify some things by removing parameters that Rare doesn't need at all.
- Refactored the bloom filter to track a Set of murmur hashes, then covert over to a Bloom after a certain threshold
- Track buckets as they are created so we can trip `max_buckets`.
- Track approximate memory usage in CircuitBreaker
- More tests, comments

I think this is probably ready for another round of reviews.

colings86 reviewed (commented):
@polyfractal I added some comments but I didn't get all the way through reviewing yet

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Rework rare_terms to use CuckooFilter

Implements a scaling cuckoo filter instead of bloom filter

Commit:
Update documentation

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Remove benchmark, temp was not supposed to be committed

polyfractal commented:
@elasticmachine test this please

polyfractal commented:
I think this is ready for the next round of reviews.  Notable changes:

- Bloom filters are out, CuckooFilters are in. :bird:   
- CuckooFilters reject inserts once they saturate (unlike blooms) so we know immediately when the filter is full.  To scale, there is a `SetBackedScalingCuckooFilter` abstraction that sits on top of `CuckooFilter`.  This maintains a list of filters and automatically adds new ones when the "current" filter is full.  It also maintains an exact set under some `threshold`, and converts to Cuckoo approximation after that threshold is reached

The original paper doesn't specify how or if cuckoo filters can merge, but I worked up an implementation that I think works.  Since all filters are created with the same parameters, a specific item can only land in two different buckets.  So we can check if the fingerprint is in the appropriate buckets across all the different filters.  If it doesn't exist, we try to insert the fingerprint using the normal insertion process.  This process might fail and we create an entirely new filter to house the fingerprint in the worst case scenario.

The alternative is to just append all the filters to one big list.  This is obviously faster, but less compact.  From some informal testing it appears the optimal-merging strategy is 50% smaller, and 5-25% slower depending.  There are still some performance issues I'm chasing, but those are related to cleaning up deleted ordinals more than the merge algo.  I'm inclined to stick with the slower merging to get a better memory footprint, since RareTerms will definitely be one of the more expensive aggs that we have.

Note: this implementation uses filters that are a power of 2 so that the xor works correctly.  Other implementations out there (Guava) figured out a different fingerprint algorithm to avoid the power of 2.  Given that we scale these anyway I don't think the more complicated fingerprint algo is needed just to avoid powers of 2.

![image](https://user-images.githubusercontent.com/1224228/55419408-b1297f00-5542-11e9-8a3a-4c301c00bcd0.png)

Accuracy testing shows that precisions of `0.001` give 0-1% error on 80m docs (10m distinct values, ranging from 10-100k "rare" terms).  Quick'n'dirty test but shows that the agg is in the realm of correct :)

![image](https://user-images.githubusercontent.com/1224228/55419401-abcc3480-5542-11e9-98fa-fc3e601911f5.png)


Note: in this case, "inaccuracy" means omitting a rare term when it should exist.  Due to the algorithm, we only have false-negatives, never false-positives (a term that is not "rare" being marked as "rare")

colings86 commented on docs/reference/aggregations/bucket/rare-terms-aggregation.asciidoc:
```diff
@@ -0,0 +1,291 @@
+==== Maximum document count
+
+The `max_doc_count` parameter is used to control the upper bound of document counts that a term can have.  There
+is not a size limitation on the `rare_terms` agg like `terms` agg has.  This means that _all_ terms
```
```suggestion
is not a size limitation on the `rare_terms` agg like `terms` agg has.  This means that terms
```
The `_all_` here makes it seem that this is exact which it's not. I know we explain this below but I think we should avoid confusion here

colings86 commented on docs/reference/aggregations/bucket/rare-terms-aggregation.asciidoc:
```diff
@@ -0,0 +1,291 @@
+1. Values are placed in a map the first time they are seen.
+2. Each addition occurrence of the term increments a counter in the map
+3. If the counter > the `max_doc_count` threshold, the term is removed from the map and placed in a CuckooFilter
+4. The CuckooFilteris consulted on each term.  If the value is inside the filter, it is known to be above the
```
```suggestion
4. The CuckooFilter is consulted on each term.  If the value is inside the filter, it is known to be above the
```

colings86 commented on docs/reference/aggregations/bucket/rare-terms-aggregation.asciidoc:
```diff
@@ -0,0 +1,291 @@
+
+1. Values are placed in a map the first time they are seen.
+2. Each addition occurrence of the term increments a counter in the map
+3. If the counter > the `max_doc_count` threshold, the term is removed from the map and placed in a CuckooFilter
```
Should we link to the CuckooFilter paper here?

colings86 commented on server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java:
```diff
@@ -0,0 +1,474 @@
+ * NOTE: this version does not support deletions, and as such does not save duplicate
+ * fingerprints (e.g. when inserting, if the fingerprint is already present in the
+ * candidate buckets, it is not inserted).  By not saving duplicates, the CuckooFilter
+ * loses the ability to delete values.
```
Maybe we should have a link to the paper here?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractRareTermsAggregator.java:
```diff
@@ -0,0 +1,141 @@
+                                DocValueFormat format, T valuesSource, U includeExclude) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+
+        this.filter = new SetBackedScalingCuckooFilter(10000, Randomness.get(), precision);
```
One potential issue I see here is that the `Random` we get from `Randomness` will have a different seed each time meaning that multiple identical requests are not guaranteed to have the same response. I wonder if we should use a seed to overcome this? Maybe the index_shardID would work, or maybe there is something from the request we could use for a seed?

polyfractal replied:
Mmm, good point.  Will poke around

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractRareTermsAggregator.java:
```diff
@@ -0,0 +1,141 @@
+             */
+            throw new IllegalStateException("RareTerms agg [" + name() + "] is the child of the nested agg [" + nestedAgg
+                + "], and also has a scoring child agg [" + scoringAgg + "].  This combination is not supported because " +
+                "it requires executing in [depth_first] mode, which the RareTerms agg cannot do.");
```
Maybe we should mention this limitation in the docs?

colings86 reviewed (dismissed):
@polyfractal I took a pass and left some comments. I have not looked at the cuckoo filter implementation yet nor `SetBackedScalingCuckooFilter` as I wanted to take a look at them separately in isolation. I intend to look at these soon though.

colings86 committed:
Apply suggestions from code review

Co-Authored-By: polyfractal <polyfractal@elastic.co>

Commit:
Address review comments

- Docs tweaks
- Comments
- Remove map used by each aggregator and rely on ords instead,
which allowed refactoring part of the logic to the abstract class

polyfractal commented:
Pushed a commit addressing the review comments.  In particular, you were right about being able to remove the extra map inside each aggregator.  The only thing that turned out a bit hacky is that we need to do one extra collection of values when they are over-threshold, so that the doc_count is updated and correctly over-threshold.  This is needed later when GCing and building the aggregation since we don't have a map to check counts on anymore.

We could make the `docCounts` protected and update it directly (instead of collecting), but the extra collection seemed like a better alternative since we know it will be deferred anyway... and it doesn't escalate visibility of something like `docCounts`

A nice side effect is by removing the map, it became easier to refactor the logic into the abstract class (with the help of some accessory abstract methods)

colings86 commented on server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java:
```diff
@@ -0,0 +1,481 @@
+        // If we get this far, we failed to insert the value after MAX_EVICTION rounds,
+        // so cache the last evicted value (so we don't lose it) and signal we failed
+        evictedFingerprint = fingerprint;
+        return false;
```
It looks like if we add an entry when the filter is saturated that it will still change the filter since it still runs and maintains the evictions it tries while it works out that the filter is full? I'm wondering if this property of the filter is intended and further if its considered ok?

colings86 replied:
Oh actually this might be ok because of https://github.com/elastic/elasticsearch/pull/35718/files#diff-53984b5b7b1ac2bfe9cafb420e574923R196 which checks the evicted fingerprint. Is that intended to overcome the problem above?

polyfractal replied:
We chatted about this offline, but for posterity, the second comment is correct.  

After an item runs through a full eviction cycle, if we can't find a slot we need to save the item to a holding-area before returning false.  If we don't, we'll have accidentally kicked an item out and lose it.  We could theoretically put it back into the map by "undoing" the evictions... but that would be relatively expensive compared to a dedicated eviction variable.

colings86 reviewed (commented):
I did a review pass on CuckooFilter and left some comments

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
```
I wonder if this method should return a new instance? We had problems in the past when we reduced shard results into the first shards results rather than reducing them into a new instance and I wonder if we might trip up on the same kinds of problems again here?

polyfractal replied:
Hmm, that could get very expensive since merge is called repeatedly for each new shard result?

The datastructure doesn't offer this protection itself, but in [InternalMappedRareTerms](https://github.com/elastic/elasticsearch/pull/35718/files/d9998d59d6c368f9ed4f57a099aa091f1ebac147#diff-e8b986b34de85409f29dd4b0fca3809dR122) we initialize a fresh `SetBackedCuckooFilter` instead of reusing one (that's why the class has a copy ctor)

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            if (hashes.size() > threshold) {
+                convert();
```
The `convert()` method assumes that `cuckoo_capacity > 2 * set_threshold` but this is not enforced anywhere in this class. I wonder if we should either enforce it (seems easiest solution to me) or make sure the `convert()` method can cope with the filter getting full during the conversion (i.e. create multiple filters during the loop)

polyfractal replied:
Yeah agreed, I can throw some exceptions and/or asserts to guarantee they are the same.  Was hardcoded so I just ignored it... but in case that changes in the future would be good to have checks here

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+        } else if (isSetMode && other.isSetMode == false) {
+            // Other is in cuckoo mode, so we convert our set to a cuckoo then merge collections.
+            // We could probably get fancy and keep our side in set-mode, but simpler to just convert
+            convert();
```
Same comment as above but the assumption here is `cuckoo_capacity > set_threshold`

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+            // Other is in cuckoo mode, so we convert our set to a cuckoo then merge collections.
+            // We could probably get fancy and keep our side in set-mode, but simpler to just convert
+            convert();
+            filters.addAll(other.filters);
```
Why do we not attempt to merge the filters in this case?

polyfractal replied:
Hmm, good point.  This was written before I worked out the proper merge.  But besides not being ideal, I think this is actually a bug.

The way we're using the filters ensures there are no duplicates anywhere... but this line could accidentally introduce duplicates.  I'll switch to a proper merge.

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+            int current = 0;
+            CuckooFilter currentFilter = filters.get(current);
+
+            for (CuckooFilter filter : other.filters) {
```
nit: for clarity could we call the variable `otherFilter`? This might avoid confusion below on whether its the filter in this instance or the `other` instance

colings86 commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+                        boolean success = false;
+
+                        // If the fingerprint is new, we try to merge it into the filter at our `current` pointer.
+                        // This might fail (e.g. the filter is full), so we may have to try multiple times
```
Since we have already discounted the case where this fingerprint already exists in one of the filters, and by virtue of the fact there are multiple filters all but the last filter must have been declared "full" before, won't the subsequent while loop likely always descend to the last filter? If so, shouldn't we simplify this by either starting `current` at `filters.size() - 1` or removing this logic and just keeping a reference to the last filter, updating this reference as we need to create more filters?

polyfractal replied:
++ I believe you're correct, good catch.  This is leftover from an earlier iteration that didn't check if the fingerprint existed.  It's not necessary now.

colings86 reviewed (commented):
@polyfractal I left some comments on the merge method of `SetBackedScalingCuckooFilter`

polyfractal commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+     */
+    public void registerBreaker(Consumer<Long> breaker) {
+        this.breaker = breaker;
+        breaker.accept(getSizeInBytes());
```
The breaker defaults to a no-op Consumer: https://github.com/elastic/elasticsearch/pull/35718/files/d9998d59d6c368f9ed4f57a099aa091f1ebac147#diff-fcea5bc0639003e5f340598671d7a8e7R65, but that does assume no one accidentally uses the setter with `null`.  I'll add a null check there.

The reason for this strange arrangement is because we don't always have a breaker available.  E.g. the breaker is available in the aggregators, but when the result is serialized over the wire, we no longer have a breaker because were "outside" the normal aggregator life cycle when doing the reductions.

I might be missing something, but I didn't see a way to grab the breaker service after serialization to the coordinating node, in the Internal* result classes.

polyfractal commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,361 @@
+     * If we still holding values in a set, convert this filter into an approximate, cuckoo-backed filter.
+     * This will create a list of CuckooFilters, and null out the set of hashes
+     */
+    private void convert() {
```
:+1:

Commit:
Address review comments

polyfractal commented:
Pushed a commit with most of the review comments addressed

colings86 commented on server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java:
```diff
@@ -95,6 +97,27 @@
+            throw new IllegalArgumentException("Attempted to create [" + numBuckets * entriesPerBucket
+                + "] entries which is > Integer.MAX_VALUE");
+        }
+        // TODO this is probably super slow, but just used for testing atm
```
Maybe this warning should be in aJavaDoc on this constructor since it indicates this constructor should only be used in tests?

polyfractal replied:
:+1:

colings86 reviewed (commented):
I left a comment but this is looking good to me. Since this will introduce an entirely new algorithm with some complex logic for making it scalable I wonder if @jpountz would like another look before we merge it?

polyfractal commented:
Thanks @colings86 :)  Since the code is settling down a bit I'm going to run a few more tests and generate images to include in the docs, similar to what we have for the other approximate aggs.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractRareTermsAggregator.java:
```diff
@@ -0,0 +1,200 @@
+                    }
+                }
+            }
+        }
```
Note to self: we may want to check if the value exists in the map here.  E.g. value is in map (sub threshold), but cuckoo throws false positive so we don't check/increment the value in the map which could lead to a "rare" term that shouldn't be.

Commit:
Performance improvements

- Restructures the algorithm so that the sketch is only built at the very
end.  This means it collects all the counts/ordinals like the terms
agg (in deferred mode), then prunes any counts over-threshold and adds
them to the sketch.  This means there is more runtime overhead, but
drastically less computation.  It still maintains sub-linear size
when serializing to coordinating node.  Comparable performance to
terms agg.

- Removes java streams in hot code, extra allocations really kill
performance
- Adds a lighter 64bit murmur hash for longs.  Still uses 128bit algo
for BytesRef
- Stores just the needed 64 bits of the hash, instead of the full 128bit
Murmur3 hash.
- Exposes a number of expert APIs so the Scaling filter implementation
can calculate hashes/buckets once for all the lookups

polyfractal commented:
Keep getting pulled into other matters, but finished up some accuracy testing.  Test was a single shard with `n` distinct values (from 1m to 100m) per shard, and `k` of those values being "rare (from 1 to 100k).  Rare meaning `doc_count == 1`, where the rest had `doc_count == 6`.

That means the size of indices are `n * 6`, so the 100m test was an index with 600m docs, etc etc.

![image](https://user-images.githubusercontent.com/1224228/58965369-7fb97380-877e-11e9-8952-b095c4c6e1e8.png)
![image](https://user-images.githubusercontent.com/1224228/58965373-847e2780-877e-11e9-8107-569073905e75.png)
![image](https://user-images.githubusercontent.com/1224228/58965382-87791800-877e-11e9-9935-24acd389268f.png)
![image](https://user-images.githubusercontent.com/1224228/58965404-9233ad00-877e-11e9-8f14-76110c7a3ce1.png)


I'm fairly pleased with the results.  `precision: 0.0001` gives about 1-3% error in most cases, and `precision: 0.00001` is <= 0.1% in all tested scenarios.  The coarser precisions are acceptable if cardinality of the field is low, but fall apart quickly as cardinality rises.

Looking at the charts, it's pretty clear that the agg gets more "accurate" as the number of rare terms increases.  This is a function of inverting the usage of the CuckooFilter, turning the filter's false-positives into false-negatives.  When there is only one or ten rare values missing one of those drastically skews the error rate.  But when there are 100k, missing a few has a lower impact on the overall error rate.

polyfractal commented:
Right, did some more rigorous testing after @colings86 noticed the odd jump to 40% error in the above tests.  I think something got messed up in the test index.  Wiped the node, and rebuilt the tests to be more exhaustive and rigorous.  I've spot checked a bunch of these results and think they are (mostly) legitimate now.

### Test Setup
- Single shard
- Indexed `n` distinct terms (`n == 1-20 million`), where `k` of those terms are "rare" (`k == 1-100k`) 
- Tested precisions `0.01`, `0.001`, `0.001`
- Re-ran the test 10 times per test tuple, using a script to randomize the values (but not rarity). This helps defeat hash collision artifacts
- Also collected memory stats at every 100k values inserted into the filter

#### Precision 0.01 
![image](https://user-images.githubusercontent.com/1224228/59445716-1cf35800-8dce-11e9-9cb1-9bd21c26cd43.png)

Summary: < 14% error.  The 100% error spikes are for the single rare term, meaning it didn't find the one rare term.

#### Precision 0.001 
![image](https://user-images.githubusercontent.com/1224228/59445762-2bda0a80-8dce-11e9-90cb-901ca5d299ca.png)

Summary: < 2.5% error.  You can see how lower rare values (1, 100) have more discrete steps than higher values (1000, 100k).  I think this is a function of missing a value having a larger impact on error (1 out of 10 is much bigger than 1 out of 100k), which makes the chart less smooth.

#### Precision 0.0001 
![image](https://user-images.githubusercontent.com/1224228/59445777-34cadc00-8dce-11e9-9118-f88d7618aaa1.png)

Summary: < 0.4%, with two spikes to 1%.  The 1% errors on 100 (e.g. 1 out of 100 missing) is interesting, and I don't have an explanation.  Might just be happenstance, or something else.  Will followup.

#### Memory usage
![image](https://user-images.githubusercontent.com/1224228/59446191-0699cc00-8dcf-11e9-8590-7fff1aad241c.png)

Summary: we can see how memory increases roughly every 1 million values (which is the theoretical capacity we define for each `CuckooFilter`).  It's not always 1m since there is some approximation there depending on configured precision.

On one hand I like how predictably linear this is, at all precisions.  Makes it easy to say "if you have 20m unique values at precision 0.001, it'll use about 30mb".

OTOH, memory usage is higher than I'd like.  It will definitely be one of heavier (maybe heaviest?) aggregations we have.  I think the saving grace here is that it will hopefully be used in situations which limit unique number of terms, e.g. you want to know rare IPs in the last day/week/month, not the last ten years.

Also for reference, to do this accurately with terms agg, 20m buckets would be `20m * 69b == ~1.38gb` (with `69b` being a very optimistic estimate of an empty bucket cost, based on a number of `long` and a few empty object headers)

### Future

Based on the graphs, I think we should set the default precision to `0.001`.  It has a nice accuracy profile (`(n *  10⁻⁹)%`, e.g. 20m distinct terms ~== 2% error), and uses the same memory as `0.01`.

I think the only variable I haven't tweaked yet is the initial capacity we set the filter, hardcoded to 1m right now.  I suspect larger filters will improve accuracy (due to having fewer filters), increase the size of each filter and the "duration" of the memory step function.  It might also have a small positive improvement on runtime, since there are fewer filters to check and fewer hashes to calculate.  I wouldn't expect it to be significant though.

I also haven't tested intermediate precisions.  E.g. all the precisions are orders of magnitude of 0.1, but you could configure say 0.5 or 0.003.  I think these should interpolate well inbetween the existing graphs though, don't see it changing things dramastically.

As always, open to any comments/questions/concerns :)

colings86 commented:
> Based on the graphs, I think we should set the default precision to 0.001. It has a nice accuracy profile ((n * 10⁻⁹)%, e.g. 20m distinct terms ~== 2% error), and uses the same memory as 0.01.

Do we know why `0.001` and `0.01` precisions use the same memory? I don't think this necessarily indicates a problem, I'm just curious

polyfractal commented:
It's an interesting effect :)  If you run the numbers:

#### 0.01
- 2 entries per bucket
- load factor of 84%
- 9 bits per entry. 
- For a 1m capacity, it needs 595238 buckets.  Next power of 2 is 1048576 buckets. 

`1048576 buckets * 2 entriesPerBucket * 9 bitsPerEntry == 2.359mb` per filter

#### 0.001 
- 4 entries per bucket
- load factor of 95.5%
- 13 bits per entry
- For a 1m capacity, it needs 261780 buckets.  Next power of 2 is 262144 buckets 

`262144 buckets * 4 entriesPerBucket * 12 bitsPerEntry == 1.572 mb` per filter


(These theoretical numbers are a bit under what the code is reporting, but there is some additional overhead like the set, internal bookkeeping of PackedInts, etc.  They largely align with the actual values)

The power of 2 requirement over-inflates the `0.01` case and nearly doubles the number of buckets, which means it absorbs ~1.35 million values on average before saturating, whereas the `0.001` case is very close to the theoretical ~1m values before saturating.  

So basically the `0.01` case has a larger filter due to power of two condition, but that also means it can absorb more values (even at a lower load factor) which means the two precisions end up near the same memory overhead.

At least, that's my interpretation :)

Commit:
Update defaults, add charts to docs

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Compile errors, checkstyle

Commit:
Add blurb to docs about max_buckets

Commit:
Fix test

Commit:
Update comments

polyfractal commented:
@colings86 Pushed some updates to the documentation and tidied up some tests/comments.  I think the new algo changes are ok to review. :crossed_fingers:

colings86 commented on docs/reference/aggregations/bucket/rare-terms-aggregation.asciidoc:
```diff
@@ -174,7 +174,23 @@ which match the `max_doc_count` criteria will be returned.  The aggregation func
+To limit the danger of this setting, the maximum `max_doc_count` is 100.
+
+[[search-aggregations-bucket-rare-terms-aggregation-max-buckets]]
+==== `search.max_buckets`
```
Nit: I'm not a fan of heading that are settings names, could we maybe call this something like "Max bucket limit"?

colings86 commented on server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java:
```diff
@@ -52,6 +50,10 @@
  *
+ * NOTE: this CuckooFilter exposes a number of Expert APIs which assume the caller has
+ * intimate knowledge about how the algorithm works.  It is recommended to avoid these
+ * APIs, or better yet, use {@link SetBackedScalingCuckooFilter} instead.
```
I'd just say its recommended to use `SetBackedScalingCuckooFilter` and not touch this class

colings86 commented on server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java:
```diff
@@ -477,4 +518,13 @@ public boolean equals(Object other) {
             && Objects.equals(this.evictedFingerprint, that.evictedFingerprint);
     }
+
+    static long murmur64(long h) {
```
Should we add this to the MurmurHash3 class?

colings86 reviewed (commented):
Did a really quick pass but I need to more thoroughly go through the CuckooFilters again

iverase commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,409 @@
+            // Rather than converting the other to a cuckoo first, we can just
+            // replay the values directly into our filter.
+            other.hashes.forEach(this::add);
+            maybeConvert();
```
Is maybeConvert() necessary?

polyfractal replied:
Ah no, good catch.  I don't think this is needed.

iverase reviewed (commented):
I had a look into CuckooFilter and SetBackedScalingCukooFilter, very cool. I left some comments in the merging logic.

polyfractal commented on server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java:
```diff
@@ -0,0 +1,409 @@
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            maybeConvert();
```
Hmm, yeah, we can go about twice over the threshold.  Tricky to estimate if we should convert to a filter first though.  If both sets are duplicates of each other, the total size might not change (or change much).  But we won't know that until we've merged them together.

I think it won't matter too much if we go twice over the threshold, since the threshold is set very low relative to the size of the filters.  E.g. the current (hard coded) threshold is 10,000 hashes.  So 20k longs would be ~ 160kb, compared to the initial filter size of ~1.7mb.

iverase replied:
that works for me

Commit:
Merge remote-tracking branch 'origin/master' into rare_terms3

Commit:
Address review comments, merge conflicts

Commit:
Skip segments when there are no buckets after merging

polyfractal commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rare_terms3

Commit:
Add version skip to yaml tests

iverase commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalRareTerms.java:
```diff
@@ -0,0 +1,202 @@
+            B read(StreamInput in, DocValueFormat format) throws IOException;
+        }
+
+        long bucketOrd;
```
@polyfractal could you comment here, I don't see how this is used and it is not part of equal or hashCode?

polyfractal replied:
@iverase  Right thanks, forgot to comment on this.  So this is a slightly strange idiom that the terms agg uses, and rare_terms just inherited the same behavior.

The `bucketOrd` on each bucket is a temporary "scratch" space for noting down the bucket ordinal at runtime, which is used to run deferred aggs after collecting the relevant buckets.  But it's not serialized between nodes, and not part of the equals/hashcode/etc.

E.g. when we are building the sketch, we skip buckets that are over-threshold and [re-assign ordinals of candidate buckets](https://github.com/elastic/elasticsearch/pull/35718/files#diff-29eed13231a9e57022105c0d193c398cR120).  The new ordinals are then used for [deferred aggregations](https://github.com/elastic/elasticsearch/pull/35718/files#diff-29eed13231a9e57022105c0d193c398cR149) after we merge out the empty buckets.

Regular terms have a [similar `bucketOrd](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java#L60) and uses it as a temporary scratch space to note down ordinals when running the [top-n priority queue](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTermsAggregator.java#L144)

It's not a great pattern and it'd be nice if we could remove it, but I think we should make the change to all the terms/sigterms/rareterms at the same time since it's a pattern used across all of them.

iverase replied:
> I think we should make the change to all the terms/sigterms/rareterms at the same time since it's a pattern used across all of them.

I agree and if we cannot remove it at least rename it to something that suggest is a scratch value :)

Commit:
Remove merge conflict .orig file

iverase reviewed (approved):
LGTM

polyfractal commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

polyfractal commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rare_terms3
